### PR TITLE
[Enhancement] Restrict internal response body types

### LINKED ISSUE #408: Refactor: move extract functions to `io.ReadCloser` or `io.Reader`
Currently `body` argument has too wide type `interface{}`, is should be changed to `io.Reader` or `io.ReadCloser` to simplify extract usage

### PR DESCRIPTION
**Pull request by outcatcher**
### What this PR does / why we need it
Restrict `extract` method arguments to `io.Reader`

Make `Result.Body` a byte slice

Make pagination handling less cryptic

### Which issue this PR fixes

Resolves #408

### Special notes for your reviewer
The reason why `Result.Body` is `[]bytes` - we can read and unmarshall it multiple times. Readers are expected to be read once.

#### Progress:

- [x] lint passing
- [x] pagination unit tests passing
- [x] other unit tests passing (here previous implementation iterations died)
- [x] acceptance tests passing

### CODE REVIEW

**Comment by outcatcher** [comment]
Note: `json.NewDecoder(reader).Decode(to)` not failing if there is invalid JSON after closing root element, but current approach with `io.ReadAll` does, e.g with:
```json
{
  "element": {}
}
}
```

That's the  reason of change is some unit tests fixtures

**Comment by outcatcher** [comment]
I suppose, it can be marked ready, as I need approvals for the `post` 🙂

**Comment by Aloento** [comment]
What are the requirements for `post` can be run?

**Comment by outcatcher** [comment]
> What are the requirements for `post` can be run?

At least one approval, AFAIK

**Comment by vladimirvshivkov** [comment]
At least two approvals

File: internal/extract/json_test.go
Comment (Aloento) [inline internal/extract/json_test.go:26]: What is the purpose of this?

**Review by Aloento — APPROVED** [review]

**Review by Aloento — APPROVED** [review]

File: internal/extract/json_test.go
Comment (outcatcher) [reply internal/extract/json_test.go:26]: Testing the case with `io.ReadCloser` passed to extract functions.

**Review by Aloento — DISMISSED** [review]

**Comment by Aloento** [comment]
acceptance/openstack/waf/v1 `FAIL: TestCertificateLifecycle (0.93s)`
```
Failure in certificates_test.go, line 52: unexpected error "error converting page body to slice: error extracting 
{
   "total": 1,
   "items": [
      {
         "id": "d2654517ae5e4cb7a7a55c2fe5594734",
         "name": "test_cert-quP3l_updated",
         "expireTime": 1653989152000,
         "timestamp": 1662396802060
      }
   ]
}
into *[]interface {}: json: cannot unmarshal object into Go value of type []interface {}"
```

**Review by Aloento — DISMISSED** [review]

**Comment by Aloento** [comment]
acceptance/openstack/waf/v1 TestCertificateLifecycle

```
certificates_test.go:55: Failure in certificates_test.go, line 55: unexpected error "error extracting 
[{
	"expireTime": 1653989152000,
	"id": "f7e44f8c908b48ca8da23c75797fb11b",
	"name": "test_cert-qJB60_updated",
	"timestamp": 1662411479809
}]
into *map[string]interface {}: json: cannot unmarshal array into Go value of type map[string]interface {}"
```

**Review by Aloento — APPROVED** [review]

**Comment by Aloento** [comment]
acceptance/openstack/cts/v1 TestTrackersLifecycle

```
Failure in helpers.go, line 53: 
unexpected error "obs: service returned error: Status=400 Bad Request, Code=MalformedXML, Message=The XML you provided was not well-formed or did not validate against our published schema, RequestId=0000018311EF90014DCCACA6A69FA18D"
```

Funny, I passed this test with this pr.

```
=== RUN   TestTrackersLifecycle
    helpers.go:15: Attempting to create OBS bucket
    helpers.go:28: Created OBS Bucket: obs-cts-testi9olo
    helpers.go:62: Attempting to create SMNv2 topic
    helpers.go:75: Created SMNv2 Topic: urn:smn:eu-de:180db9ffe90d46f8b76fae9ba0cd8eba:smn-cts-testtbu7f
    trackers_test.go:25: Attempting to create CTSv1 Tracker
    trackers_test.go:42: Created CTSv1 Tracker: system
    trackers_test.go:47: Attempting to update CTSv1 Tracker: system
    trackers_test.go:54: Updated CTSv1 Tracker: system
    trackers_test.go:36: Attempting to delete CTSv1 Tracker: system
    trackers_test.go:39: Deleted CTSv1 Tracker: system
    helpers.go:80: Attempting to delete SMNv2 topic: urn:smn:eu-de:180db9ffe90d46f8b76fae9ba0cd8eba:smn-cts-testtbu7f
    helpers.go:85: Deleted SMNv2 Topic: urn:smn:eu-de:180db9ffe90d46f8b76fae9ba0cd8eba:smn-cts-testtbu7f
    helpers.go:34: Attempting to delete OBS bucket: obs-cts-testi9olo
    helpers.go:54: Deleted OBS Bucket objects: [{{ } CloudTraces/ }]
    helpers.go:58: Deleted OBS Bucket: obs-cts-testi9olo
--- PASS: TestTrackersLifecycle (5.79s)
PASS
```

@anton-sidelnikov can you check?

image

**Comment by outcatcher** [comment]
Acceptance tests are passing now. Waiting for the second approval.
@anton-sidelnikov @artem-lifshits @vladimirvshivkov

**Review by artem-lifshits — APPROVED** [review]

**Pull request merged**
